### PR TITLE
Platform-specific typography

### PR DESCRIPTION
**Pull request by cbracken**

### CODE REVIEW

File: packages/flutter/lib/src/material/typography.dart [hunk 1]
```diff
@@ -269,11 +296,27 @@ class TextTheme {
...
+  final TextTheme black;
 
   /// A material design text theme with light glyphs.
-  static const TextTheme white = const TextTheme._white();
+  final TextTheme white;
+
+  const Typography._(this.black, this.white);
+
+  /// Returns the default typography for the specified platform.
+  static Typography forPlatform(TargetPlatform platform) {
```
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: Thoughts on static `forPlatform` vs a factory ctor with required `TargetPlatform`? The only precedent I could find for a sort-of similar platform-specific ctor was the scroll behaviour delegate, but that's a type that actually gets used as a delegate as opposed to this case, which is just a util lookup.

Another option would be to just add black/white getters.

**Comment by cbracken** [comment]
Rebased this on top of #6640 to get the tests happy.

File: dev/manual_tests/card_collection.dart [hunk 2]
```diff
@@ -358,6 +356,9 @@ class CardCollectionState extends State<CardCollection> {
     if (_dismissDirection == DismissDirection.endToStart)
       rightArrowIcon = new Opacity(opacity: 0.1, child: rightArrowIcon);
 
+    ThemeData theme = Theme.of(context);
```
Comment (Hixie) [inline dev/manual_tests/card_collection.dart]: these can be final

File: dev/manual_tests/raw_keyboard.dart [hunk 3]
```diff
@@ -41,6 +41,7 @@ class _HardwareKeyDemoState extends State<RawKeyboardDemo> {
 
   @override
   Widget build(BuildContext context)  {
+    TextTheme textTheme = Theme.of(context).textTheme;
```
Comment (Hixie) [inline dev/manual_tests/raw_keyboard.dart]: these can be final

File: packages/flutter/lib/src/material/typography.dart [hunk 4]
```diff
@@ -77,6 +78,32 @@ class TextTheme {
       caption  = const TextStyle(fontFamily: 'Roboto', inherit: false, fontSize:  12.0, fontWeight: FontWeight.w400, color: Colors.white70, textBaseline: TextBaseline.alphabetic),
       button   = const TextStyle(fontFamily: 'Roboto', inherit: false, fontSize:  14.0, fontWeight: FontWeight.w500, color: Colors.white,   textBaseline: TextBaseline.alphabetic);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: rename _black and _white as well.

Maybe blackMountainView and blackCupertino etc.

File: packages/flutter/lib/src/material/typography.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: Ah, yes. I think I prefer the constructor.

Also, constructor before fields please.

Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: (the rule is "code should be in the order it's run", to make it easier to read. The constructors are run before you do anything with fields, hence they go first.)

File: dev/manual_tests/card_collection.dart (hunk 2 above)
Comment (cbracken) [inline dev/manual_tests/card_collection.dart]: Done.

File: dev/manual_tests/raw_keyboard.dart (hunk 3 above)
Comment (cbracken) [inline dev/manual_tests/raw_keyboard.dart]: Done.

File: packages/flutter/lib/src/material/typography.dart (hunk 4 above)
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: I like it. Done.

**Comment by cbracken** [comment]
Comments addressed and re-squashed. PTAL.

File: packages/flutter/lib/src/material/typography.dart [hunk 5]
```diff
@@ -269,11 +297,33 @@ class TextTheme {
 ///  * [ThemeData]
 ///  * <http://material.google.com/style/typography.html>
 class Typography {
-  Typography._();
+  const Typography._(this.black, this.white);
+
+  /// Creates the default typography for the specified platform.
+  factory Typography({@required TargetPlatform platform}) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: style nit: spaces around the arguments please.

File: packages/flutter/lib/src/material/typography.dart [hunk 6]
```diff
@@ -269,11 +297,33 @@ class TextTheme {
...
+        break;
+      case TargetPlatform.iOS:
+        typography = const Typography._(
+            const TextTheme._blackCupertino(),
+            const TextTheme._whiteCupertino(),
+        );
+        break;
+    }
+    assert(typography != null);
+    return typography;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: then you'd just return null here.

File: packages/flutter/lib/src/material/typography.dart [hunk 7]
```diff
@@ -269,11 +297,33 @@ class TextTheme {
 ///  * [ThemeData]
 ///  * <http://material.google.com/style/typography.html>
 class Typography {
-  Typography._();
+  const Typography._(this.black, this.white);
+
+  /// Creates the default typography for the specified platform.
+  factory Typography({@required TargetPlatform platform}) {
+    Typography typography;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: assert platform != null

File: packages/flutter/lib/src/material/typography.dart [hunk 8]
```diff
@@ -255,10 +283,10 @@ class TextTheme {
 
 /// The two material design text themes.
 ///
-/// [Typography.black] and [Typography.white] define the two text themes used in
-/// material design. The black text theme, which uses dark glyphs, is used on
-/// light backgrounds in light themes. The white text theme, which uses light
-/// glyphs, is used in dark themes and on dark backgrounds in in light themes.
+/// [black] and [white] define the two text themes used in material design. The
```
Comment (Hixie) [inline packages/flutter/lib/src/material/typography.dart]: sentences should start with a capital letter. To make that work here, you can write it as something like `The [black] and [white] properties provide...`

**Comment by Hixie** [comment]
Ideally this would have some sort of test.

LGTM.

File: packages/flutter/lib/src/material/typography.dart (hunk 5 above)
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: Done.

File: packages/flutter/lib/src/material/typography.dart (hunk 6 above)
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: Done.

File: packages/flutter/lib/src/material/typography.dart (hunk 7 above)
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: Done.

File: packages/flutter/lib/src/material/typography.dart (hunk 8 above)
Comment (cbracken) [inline packages/flutter/lib/src/material/typography.dart]: Re-worded.

**Comment by cbracken** [comment]
Added a minimal test to ensure that typography is defined for all target platforms. Not a ton more we can test without getting into testing constants territory.

**Comment by cbracken** [comment]
Added tests that ThemeData defaults to platform-specific font family and that ensure that we select the appropriate font family according to Apple's typography guidelines.

**Comment by cbracken** [comment]
Fixes #4956

**Pull request merged**